mainchain-follower/db-sync-follower/src/native_token/mod.rs

```diff
@@ -0,0 +1,60 @@
+use crate::db_model::{Address, NativeTokenAmount, SlotNumber};
+use crate::metrics::McFollowerMetrics;
+use crate::observed_async_trait;
+use async_trait::async_trait;
+use main_chain_follower_api::{NativeTokenManagementDataSource, Result};
+use sidechain_domain::*;
+use sqlx::PgPool;
+
+#[cfg(test)]
+mod tests;
+
+pub struct NativeTokenManagementDataSourceImpl {
+	pub pool: PgPool,
+	pub metrics_opt: Option<McFollowerMetrics>,
+}
+
+observed_async_trait!(
+impl NativeTokenManagementDataSource for NativeTokenManagementDataSourceImpl {
+	async fn get_total_native_token_transfer(
+		&self,
+		after_block: Option<McBlockHash>,
+		to_block: McBlockHash,
+		native_token_policy_id: PolicyId,
+		native_token_asset_name: AssetName,
+		illiquid_supply_address: MainchainAddress,
+	) -> Result<sidechain_domain::NativeTokenAmount> {
+		if after_block == Some(to_block.clone()) {
+			return Ok(NativeTokenAmount(0).into());
+		}
+
+		let after_slot = match after_block {
+			None => SlotNumber(0),
+			Some(after_block) => {
+				crate::db_model::get_block_by_hash(&self.pool, after_block)
+					.await?
+					.expect("Parent MC hash is valid")
+					.slot_no
+			},
+		};
+		let to_slot = crate::db_model::get_block_by_hash(&self.pool, to_block)
+			.await?
+			.expect("current MC hash is valid")
+			.slot_no;
```

Comment:
I think both of these queries would be way more efficient if we don't pay round-trip twice. We'll need to monitor this `get_total_native_token_transfer times in devnet.

I understand that having Option makes it messy in SQL.
Action required: monitor method times after integrated and deployed.

Reply:
I joined the two futures so there's (almost) one round trip cost now.